phase2/outline
# New & Improved Slot Controller

## Description
This PR implements a [Reactive Controller](https://lit.dev/docs/composition/controllers/) intended to allow for better control over slots in a variety of scenarios. 

This code is based off of original work in the following locations:
- The original Outline [SlottedController](https://github.com/phase2/outline/blob/next/src/components/controllers/slotted-controller.ts)
- [Shoelace](https://github.com/shoelace-style/shoelace/blob/next/src/internal/slot.ts)

Supersedes #141 

## Implementations
Currently, there are two methods we'd use when implementing the `SlotController`. 

### General slot helpers, leaving Light DOM intact
This provides all the methods described below, and Outline will not do any manipulation of Light DOM nodes. 
```typescript
slots = new SlotController(
  this, // This, the host element.
  false // Do not manipulate Light DOM nodes.
);
```

### General slot helpers, transfer Light DOM to Shadow DOM
This provides all the methods described below, and Outline will move all Light DOM nodes to the Shadow DOM.
```typescript
slots = new SlotController(
  this, // This, the host element.
  true // Move all Light DOM nodes to the Shadow DOM.
);
```

## Methods
The following methods are defined with this controller. 

#### `hasDefaultSlot()`
Private method to determine if nodes exist inside the unnamed slot. 

#### `hasNamedSlot(name: string)`
Private method to determine if a named slot called `name` exists. 

#### `test(slotName?: string | undefined)`
Public method that can be utilized to check for the existence of all slots.

#### `getSlots()`
Public method to return an array of named slot names.

#### `moveNamedSlot(slotName: string)`
Private method that will move a specific named slot into the Shadow DOM.

#### `moveDefaultSlot()`
Private method to move all the nodes inside the default slot into the Shadow DOM.

#### `moveSlots()`
Private method that iterates and handles moving both named and unnamed slot content. This calls both `moveDefaultSlot()` and `moveNamedSlot()` for each named slot. Only enabled if `shadowShift` is set as `true`. 

## Testing
- Open GitPod
- Check the Breadcrumb component
- See that it looks the same as it did before at https://outline.phase2tech.com/?path=/docs/navigation-breadcrumbs--page
- Currently also used as a replacement for the original `SlottedController` in about 12 components on another internal project. The link to that PR can be seen in the history if you have appropriate Phase2 permissions. 

<a href="https://gitpod.io/#https://github.com/phase2/outline/pull/203"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
